src/main/scala/firrtl/options/Stage.scala

```diff
@@ -44,18 +23,33 @@ abstract class Stage {
     * @param annotations input annotations
     * @return output annotations
     */
-  def execute(annotations: AnnotationSeq): AnnotationSeq
+  def run(annotations: AnnotationSeq): AnnotationSeq
+
+  /** Execute this [[Stage]] on some input annotations. Annotations will be read from any input annotation files.
+    * @param annotations input annotations
+    * @return output annotations
+    */
+  final def transform(annotations: AnnotationSeq): AnnotationSeq = {
+    val preprocessing: Seq[Phase] = Seq(
+      phases.GetIncludes,
+      phases.ConvertLegacyAnnotations,
+      phases.AddDefaults )
```

Comment:
Just like we are registering transforms, have you thought about registering phases? It also seems like we really need a dependency API for both phases and transforms.

Reply:
I had not thought about that. Thinking about it now... The aim of registering transforms is to allow transforms to expose to a `Stage` that they contain options that the stage should expose to the user. This comes about because transforms consume annotations and command line options and annotations are fungible. Further, transforms can be injected into the stage from a library and may not live it the stage's package. Hence, you need the stage to discover what options it should present to the user via registration.

By that logic, if a phase wants to consume annotations that then means that it may want to expose options via it's encapsualting stage. However, I'm not sure if (or it hadn't occured to me that) phases are something that the user should get to inject into a stage.

All that said, passes and transforms may subclass phase eventually...

So... Phases should likely be registerable.

Dependency API? Maybe you could clarify on what you're thinking. I've long thought that we need for passes/transforms/phases/stages to express how they interoperate so that some scheduler can figure out a directed graph that satisfies dependencies and generates the requested outputs at each leaf. Some generalization of input/ouptput form?

Reply:
One follow-up: the registree (the thing wants to be registered) will need to provide information about what stage it wants to be registered by. Options are then exposed backwards along any stages that form a path to the shell which the user is interacting with.

E.g., a FIRRTL registered transform should register options into FIRRTL and those are exposed in Chisel and a (future) diplomacy stage. Registration of external Chisel phases would then _not_ have their options exposed to downstream tools (FIRRTL, Yosys, HAMMER).

Maybe we should discuss this next week... I can work on some mockups.